### DISCUSSION TITLE
Library extending xunit v3 appears to be a test project - how to disable?

### DISCUSSION THREAD

**Discussion by martincostello** [xunit/xunit#3093] [Question] [answered]
I'm updating my library that routes `ITestOutputHelper` into `ILogger` to support v3 (https://github.com/martincostello/xunit-logging/pull/804) and in Visual Studio it appears as if it's a test project:

![image](https://github.com/user-attachments/assets/acc8f08d-3895-4036-89c5-dd4cb0185956)

I'm also seeing https://github.com/xunit/xunit/issues/3036, which makes it seem like something is also trying to make it be an `Exe` instead of a `Library`.

What's the right way to go about getting the library to just be a plain library and not appear to be test-related itself?

I tried playing around with a few properties but that didn't seem to get me anywhere that didn't involve some sort of compiler error (either conflicting types, missing types or a missing entrypoint):

```xml
<PropertyGroup>
  <GenerateProgramFile>false</GenerateProgramFile>
  <GenerateSelfRegisteredExtensions>false</GenerateSelfRegisteredExtensions>
  <GenerateTestingPlatformEntryPoint>false</GenerateTestingPlatformEntryPoint>
  <IsTestProject>false</IsTestProject>
  <XunitAutoGeneratedEntryPoint>false</XunitAutoGeneratedEntryPoint>
</PropertyGroup>
```

**Comment by filzrev** [comment, marked as answer]
> What's the right way to go about getting the library to just be a plain library and not appear to be test-related itself?

This behavior is caused by referencing `xunit.v3.core` package. 
It transitively referencing `Microsoft.Testing.Platform.MSBuild` package. And that include `TestContainer` project capability.
https://github.com/microsoft/testfx/blob/v3.6.4/src/Platform/Microsoft.Testing.Platform.MSBuild/buildMultiTargeting/Microsoft.Testing.Platform.MSBuild.targets#L15-L16

I thought it can be fixed by following changes.

**1. `MartinContello.Logging.XUnit.v3.csproj`**
Reference `xunit.v3.extensibility.core` package instead of `xunit.v3.core` package.

**2. MartinCostello.Logging.XUnit.v3.Tests.csproj**
Remove `<GenerateSelfRegisteredExtensions>false</GenerateSelfRegisteredExtensions>` setting.

**Reply by bradwilson** [reply]
Correct, for libraries which are extensions, `xunit.v3.extensibility.core` is the package to reference.

Referencing `xunit.v3.core` treats the project like a unit test project, including all the entry-point generation and expectation that you'll be an executable rather than a library.

**Reply by bradwilson** [reply]
Also, if you're doing this for v2, the equivalent package(s) are `xunit.extensibility.core` and/or `xunit.extensibility.execution` (depending on where the things you're trying to extend live). That split no longer exists in v3, everything has been moved into core.

**Reply by martincostello** [reply]
Thanks both - I'll try this tomorrow.

Maybe the [migration guide](https://xunit.net/docs/getting-started/v3/migration) is a little wrong? I used xunit.v3.core because that's where the docs said `ITestOutputHelper` moved to:

> The follow types have been moved to Xunit (in xunit.v3.core):
>
> - ITestOutputHelper

**Reply by bradwilson** [reply]
Maybe.

The migration guide does have a brief mention of packages that extensibility authors are expected to use, but it is sitting under [a section about how to name your extension packages](https://xunit.net/docs/getting-started/v3/migration#should-extension-authors-follow-this-same-strategy), specifically:

![image](https://github.com/user-attachments/assets/1cd6001d-fa0b-475c-8e80-5bd660867e21)

But the real problem is that the migration guide does not do nearly enough for extensibility authors, and the vast majority of the content is targeted at test authors.

It is probably still worth updating those references to "which package these come from" to use the extensibility package, so I'll do that.

**Reply by bradwilson** [reply]
It's been updated:

![image](https://github.com/user-attachments/assets/6d19a457-c63b-48b2-a88b-c8741f7b0353)

Let me know if there is anything else confusing there.

I have it on my TODO list to do more extensibility porting documentation (which will go into the same migration document, likely), based at least in part on this comment: https://github.com/xunit/xunit/discussions/3084

**Reply by martincostello** [reply]
Thanks both - using `xunit.v3.extensibility.core` instead of `xunit.v3.common` and `xunit.v3.core` resolved the issues I was having.
